docs: add cryptography documentation (#280)
canonical/craft-store: docs/explanation/cryptography.rst

.. _explanation_cryptographic-technology:

Cryptographic technology in Craft Store
=======================================

Craft Store uses cryptographic processes to send files between devices and endpoints
through the internet. It does not directly implement its own cryptography, but it does
depend on external libraries to do so.

Authentication
--------------

Craft Store uses `macaroons`_, as an authentication mechanism, which are processed by
the `macaroonbakery <https://pypi.org/project/macaroonbakery/>`_ library. This library
validates and manages macaroons as returned by stores and simplifies the inclusion of
macaroons in further requests to stores.

Credentials may additionally be stored on-disk using the `keyring
<https://pypi.org/project/keyring/>`_ library, which will use the keyring service
provided by the host operating system. If the host does not have a keyring service, they
will instead be stored in a plain text file called :file:`credentials.json` under the
application's data storage directory. A warning is issued to the terminal when this

> behavior is triggered. This behavior is available to ease the usage of Craft Store
> within CI containers, but is generally discouraged.

Reviewer: ```suggestion
behavior is triggered. This behavior is available to ease the usage of Craft Store
inside virtual machines and containers, but is generally discouraged.
```

For CI we actually ease it with the environment variable option

> .. _reference_cryptography:

Reviewer 2: ```suggestion
.. _explanation-cryptographic-technology:
```
